File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    
    d = {}
    total = {}
    
    for i, (genre,play) in enumerate(zip(genres, plays)):
        if genre not in d:
            d[genre] = [(i, play)]
        else:
            d[genre].append((i, play))
        
        if genre not in total:
            total[genre] = play
        else:
            total[genre] += play
            
    for (k, v) in sorted(total.items(), key = lambda x:x[1], reverse = True):
        for (i, play) in sorted(d[k], key = lambda x:x[1], reverse = True)[:2]:
            answer.append(i)
    
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.java
import java.util.*;

class Song {
    int idx;
    int play;

    public Song(int idx, int play) {
        this.idx = idx;
        this.play = play;
    }

    @Override
    public String toString() {
        return "Song{" +
                "idx=" + idx +
                ", play=" + play +
                '}';
    }
}

class Solution {
    public int[] solution(String[] genres, int[] plays) {
        int[] answer = {};

        Map<String, Integer> map = new HashMap<>();
        for (int i = 0; i < genres.length; i++) {
            map.put(genres[i], map.getOrDefault(genres[i], 0) + plays[i]);
        }
        List<String> genreList = new ArrayList<>(map.keySet());
        Collections.sort(genreList, (o1, o2) -> map.get(o2) - map.get(o1));

        List<Integer> ans = new ArrayList<>();
        for (String genre : genreList) {
            List<Song> songs = new ArrayList<>();

            for (int i = 0; i < genres.length; i++) {
                if (genre.equals(genres[i])) {
                    songs.add(new Song(i, plays[i]));
                }
            }
            Collections.sort(songs, (o1, o2) -> o2.play - o1.play);

            if (songs.size() == 1) {
                ans.add(songs.get(0).idx);
            } else {
                ans.add(songs.get(0).idx);
                ans.add(songs.get(1).idx);
            }
        }

        answer = ans.stream().mapToInt(i -> i).toArray();

        return answer;
    }
}